cmd/proxy/actions/app_proxy.go

```diff
@@ -18,11 +23,41 @@ func addProxyRoutes(
 	app.GET("/healthz", healthHandler)
 
 	// Download Protocol
+	// the download.Protocol and the stash.Stasher interfaces are composable
+	// in a middleware fashion. Therefore you can separate concerns
+	// by the functionality: a download.Protocol that just takes care
+	// of "go getting" things, and another Protocol that just takes care
+	// of "pooling" requests etc.
+
+	// In our case, we'd like to compose both interfaces in a particular
+	// order to ensure logical ordering of execution.
+
+	// Here's the order of an incoming request to the download.Protocol:
+
+	// 1. The downloadpool gets hit first, and manages concurrent requests
```

Comment:
I find it overly complicated, do you find it readable?

Reply:
@michalpristas I 100% agree that it's more complex. I also agree that it's *less* readable than what we currently have, but that comes with the complexity part. 

My first option was just to lump all of this code into the `download.go` file and that would have been much much harder to read because you'd have **two worker pools* and a *singleflight* all under the same struct. I don't mind returning to that option if you want? Otherwise, let me know if you have other ideas, totally happy to change things :) 

Here's what's required: 
1. 2 pools (one for download.Protocol, one for go mod download) 
2. one singleflight for the operation of "go mod download" and "storage.Save"

Reply:
I imagined something simpler e.g (and I am totally not thinking this through, just writing an idea as it goes)
- the request comes to get/list/whatever - no handling of concurrency just let it come/fail
- from request, you call X to spin up worker
- X returns a channel C which will be closed once the module is prepared in Store
- X checks store for module.v, closes C if it does
- X checks whether work for module.v is already queued
-  if not it creates a job with a channel passed into goroutine and stored in a local `map[module.v]chan`. Returns chan
- if c,ok := map[module.v]; ok; return c
- spinned up job performs go get, storage.Save, close(C)
- request waits on <-C, and then serves stored module

Reply:
@michalpristas 
1. The request does come to the handler and the handler doesn't know anything about underlying concurrency, including all the download protocols that are not concurrent (i.e. goget)
2. What is X, is it all 3 requirements together? 
3. From the rest of those points, there's a missing requirement which is separating the download.Protocol Pool and the goget/storage.Save Pool. How would you split those up? 
4. I'm kind of doing everything you're saying, but more separated out than all in one file: X. I'm happy to put them all in X if you'd like :)

Reply:
I dismissed my blocking review. if folks are ok with this design I will not block and will think this through if needed once I'm mentally capable of thinking.

Reply:
@michalpristas no I 100% agree with you. I think we shouldn't rush this PR since it adds a lot of complexity. I'll mark this as do not merge until we have a bunch of people weighing in and maybe proposing different approaches. One idea is that I can have some of the Athens peeps at GopherCon doing a group review and brainstorm some other stuff and bringing it back here. I even don't mind opening another PR (or someone else opening another PR) that does the same thing differently so we can really compare. Thank you!

Reply:
Hey if you manage to get people together and brainstorm it would be great ;-)